### ISSUE TITLE
Collapse tile loops that don't do anything.

### ISSUE THREAD

**Issue by benvanik** [google/iree#5550]
This IR:
```
  %0 = linalg.init_tensor [2, 4096] : tensor<2x4096xi32>
  %1 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%arg0, %arg1 : tensor<2x4096xi32>, tensor<2x4096xi32>) outs(%0 : tensor<2x4096xi32>) {
  ^bb0(%arg2: i32, %arg3: i32, %arg4: i32):  // no predecessors
    %2 = muli %arg2, %arg3 : i32
    linalg.yield %2 : i32
  } -> tensor<2x4096xi32>
```

Results in this set of nested loops:
```mlir
      scf.for %arg0 = %workgroup_id_y to %c2 step %workgroup_count_y {
        %3 = muli %arg0, %c-1 : index
        %4 = addi %3, %c2 : index
        %5 = cmpi slt, %c1, %4 : index
        %6 = select %5, %c1, %4 : index
        scf.for %arg1 = %workgroup_id_x to %c4096 step %workgroup_count_x {
          %7 = muli %arg1, %c-1 : index
          %8 = addi %7, %c4096 : index
          %9 = cmpi slt, %c1, %8 : index
          %10 = select %9, %c1, %8 : index
          scf.for %arg2 = %c0 to %6 step %c1 {
            %11 = addi %arg0, %arg2 : index
            scf.for %arg3 = %c0 to %10 step %c1 {
              %12 = addi %arg1, %arg3 : index
              %13 = memref.load %0[%11, %12] : memref<2x4096xi32>
              %14 = memref.load %1[%11, %12] : memref<2x4096xi32>
              %15 = muli %13, %14 : i32
              memref.store %15, %2[%11, %12] : memref<2x4096xi32>
            }
          }
        }
      }
```

Which then when converted to a CFG is just an insane amount of IR.
Since we know storage is dense we should be able to collapse all of these to a single loop (or at least two loops - one for workgroup and one for inner tiles). There may already be a pass or some configuration that does this but I'm not able to find it. A pass that linearized memrefs may be the easiest thing (and I really am hoping that exists upstream already) as we'd do that prior to tiling and hopefully not generate these extra loops in the first place.

After lowering the above to VMVX - where we only have dense linear memory - it becomes this:
```mlir
    scf.for %arg11 = %arg3 to %c2 step %arg9 {
      %3 = muli %arg11, %c-1 : index
      %4 = addi %3, %c2 : index
      %5 = cmpi slt, %c1, %4 : index
      %6 = select %5, %c1, %4 : index
      scf.for %arg12 = %arg2 to %c4096 step %arg8 {
        %7 = muli %arg12, %c-1 : index
        %8 = addi %7, %c4096 : index
        %9 = cmpi slt, %c1, %8 : index
        %10 = select %9, %c1, %8 : index
        scf.for %arg13 = %c0 to %6 step %c1 {
          %11 = addi %arg11, %arg13 : index
          %12 = muli %11, %c4096 : index
          scf.for %arg14 = %c0 to %10 step %c1 {
            %13 = addi %arg12, %arg14 : index
            %14 = addi %12, %13 : index
            %15 = muli %14, %c4 : index
            %16 = vmvx.buffer.load<%0 : !vmvx.buffer>[%15] : i32
            %17 = vmvx.buffer.load<%1 : !vmvx.buffer>[%15] : i32
            %18 = muli %16, %17 : i32
            vmvx.buffer.store<%2 : !vmvx.buffer>[%15], %18 : i32
          }
        }
      }
    }
```
Where ideally some scf folding improvements (#5547) may help, however once we are down at this level it gets harder to do the kind of wide-vector operations we need: there should be zero loops here and the whole thing should end up as just a single vector operation.

**Comment by benvanik** [comment]
@antiagainst points out this which may help with some of these loops: https://github.com/google/iree/blob/main/iree/compiler/Conversion/LinalgToSPIRV/FlattenMemRefSubspanPass.cpp

**Comment by antiagainst** [comment]
The pass will flatten n-D `memref` into 1-D ones; it won't directly collapse those loops though. If we can run the flattening pass first and then tile, it might avoid generating a loop nest from the beginning. But this may only apply to the second+ level tiling given the first level is done at dispatch region formation time. Although I think by flattening the dimensions it may help to clean up those loop nest.

**Comment by benvanik** [comment]
You're totally right that it doesn't remove the unneeded loops but it does do a fantastic job of linearizing and allows me to remove the code I had for doing so! Do you mind if I move it to `iree/codegen/Conversion/Common/`?

**Comment by antiagainst** [comment]
Sure, please go for it! (Sorry I should placed it there from the beginning. ;-P)

**Comment by benvanik** [comment]
(no worries, I feel like YAGNI is a pretty decent rule with utils like this :)

**Comment by antiagainst** [comment]
Also the pass uses affine expressions for index calculation. I saw you are using std.mul and std.add there. Switching to affine might give better canonicalizations as affine ones are better supported.

**Comment by benvanik** [comment]
Ah yeah I'm running the `LowerAffinePass` - I put yours before that. Are there any other things that may optimize affine? (particularly something that may do the scf+affine combining that may help #5547?)

**Comment by benvanik** [comment]
ohhh I think I get you - I changed to using affine.map for one of my conversions and lowered affine after that, now I get some slightly different ops that may be easier to fold into the scf.for (#5547):
```mlir
          scf.for %arg13 = %c0 to %8 step %c1 {
            %13 = addi %arg11, %arg13 : index
            %14 = muli %13, %c16384 : index
            scf.for %arg14 = %c0 to %12 step %c1 {
              %15 = addi %arg12, %arg14 : index
              %16 = muli %15, %c4 : index
              %17 = addi %16, %14 : index
              %18 = vmvx.buffer.load<%0 : !vmvx.buffer>[%17] : i32
              %19 = vmvx.buffer.load<%1 : !vmvx.buffer>[%17] : i32
              %20 = muli %18, %19 : i32
              vmvx.buffer.store<%2 : !vmvx.buffer>[%17], %20 : i32
            }
          }
```
Specifically, `%15` can be used to offset the scf.for and `%16` can be used to adjust the step count. The `%17` could probably also be folded in. I feel like there may be a transform I'm missing though that does all this already - thoughts?

**Comment by MaheshRavishankar** [comment]
So taking a step back, I am skeptical about having such canonicalization/optimizations, etc. These are all solved in compilers like LLVM by using fairly complex analysis like SCEV, etc. Adding such canonicalizations to MLIR when it is not the "end compiler" seems like an overkill for MLIR.
But of course here you are trying to use MLIR as an end compiler. So we should try not to reimplement the whole host of scalar optimizations a traditional compiler implements in MLIR again (to me thats not what MLIR is meant to do, but of course nothing stopping you from doing this).

A good story for vmvx is to actually "always vectorize". The scalar fall back is just that. Its a fall back. It would be too much of a heavy lift to implement scalar optimizations like these in MLIR. So if we hit the fallback, we just pay the cost of these extra instructions.

**Comment by benvanik** [comment]
I strongly disagree :)
The line of reasoning you have there would also call for not having any folders or canonicalizers in MLIR at all - afterall, whatever poor infra at the end of the day would take care of it right?

If we are using scf for flow control - including around hlo/tosa/linalg/flow/etc - having the most basic of transformations to simplify things in totally safe ways is just as valid as having any other folder for any other op in any other dialect. When we have loops around asynchronous operations, for example, being able to chain async handles from across loops is critical and that will never end up as LLVM IR - or if it does it's not something LLVM would be able to fix.

So taking an even further step back, this is not about tiling or vectorization: it's about loops not being optimized at _any_ level of representation in MLIR - and I really don't think you can argue that is a good thing.

**Comment by asaadaldien** [comment]
Having `VMVX` with a variable length vector will make it very easy to (always) vectorize so scalar cases will be  very rare (if ever happen) vector<1> at the tile boundaries

**Comment by benvanik** [comment]
Even in vector cases we will still have code that does scalar math - even if just for addressing (which is particularly what I want to simplify) - the vectorization is a big performance multiplier (effectively 1/(k*N) the overhead for vectors of size N) and what's left after that is a hopefully small set of indexing and loop math. To be clear this is all just optimization in the VMVX case - it's more load bearing the higher we go up the stack: being able to elide a loop can mean the difference between a synchronization point and not or an allocation and not. Looking at it here just makes it easy to see :)

**Comment by asaadaldien** [comment]
> Even in vector cases we will still have code that does scalar math - even if just for addressing (which is particularly what I want to simplify) - the vectorization is a big performance multiplier (effectively 1/(k*N) the overhead for vectors of size N) and what's left after that is a hopefully small set of indexing and loop math. To be clear this is all just optimization in the VMVX case - it's more load bearing the higher we go up the stack: being able to elide a loop can mean the difference between a synchronization point and not or an allocation and not. Looking at it here just makes it easy to see :)

Sorry, I didn't make my point clear. I wasn't arguing with or against the loop optimization. just stating the variable length vectors superiority =)

**Comment by benvanik** [comment]
For sure :)
(it'd be useful if github had priorities - maybe it's worth adding a label for that - in the VMVX case this would be a P3 - would that have helped?)

**Comment by MaheshRavishankar** [comment]
> I strongly disagree :)
> 
> If we are using scf for control flow - including around hlo/tosa/linalg/flow/etc - having the most basic of transformations to simplify things in totally safe ways is just as valid as having any other folder for any other op in any other dialect. When we have loops around asynchronous operations, for example, being able to chain async handles from across loops is critical and that will never end up as LLVM IR - or if it does it's not something LLVM would be able to fix.
> 
> So taking an even further step back, this is not about tiling or vectorization: it's about loops not being optimized at _any_ level of representation in MLIR - and I really don't think you can argue that is a good thing.
> 
> (sorry - I need a nap - this came off too strong - but I am more than a little surprised you think that scf.for should not have trivial optimizations applied)

I am saying that `scf.for` is really bottom of the stack. For me when you go to `scf.for` you have materialized all of the semantics information and then you cant recover that back. So if you want more semantics information you need a different abstraction.
In any case, its just my two cents. If someone is willing to implement those optimizations, then sure!

**Comment by benvanik** [comment]
Ah - I guess I'm thinking of scf.for as an input-dialect thing we'd be getting from HLO and TOSA as well - so improvements anywhere will help that layer too. Agreed that at this super deep level they are mostly just aesthetic, but higher up the stack they are much more load bearing.

**Comment by antiagainst** [comment]
I can certainly see your points here, Mahesh, and I cannot agree more that rebuilding everything we have in traditional compiler (of which I'd assume we are mostly talking about LLVM here) is a HUGE effort. But still I suspect there will be the desire/needs/requirements to build them piece by piece gradually. There are worlds where we cannot really sink to LLVM (e.g., SPIR-V ;-P) for everything, for good or bad, which is just the reality. Even on the LLVM side, I think there is the grand goal that someday we can probably shift more of LLVM to be based on MLIR, and I'd assume it depends on this too.

Again, not arguing that we should prefer this and I'd certainly agree we'd avoid when possible. Just wanted to say that sometimes the whole system is split at different levels for various reasons, so it's more nuanced. So probably case by case. IIUC, what Ben wants here can be achieved by simple induction variable range recalculation and substitution given from the IR the induction variable is only used as a linear chain of ops and finally for indexing buffers? So the transformation is probably not that scary. :)

**Comment by benvanik** [comment]
yeah thinking mostly "is the induction variable used once for simple arithmetic? if so, fold that arithmetic into the range" - that's what most of these loops (and mhlo.while/etc) end up becoming
